battlepass/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (C) 2018-2022 GameDAO AG.
+// SPDX-License-Identifier: Apache-2.0
+
+//! BATTLEPASS
+//! This pallet provides functionality to create, manage and participate in battlepasses.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+use sp_std::convert::TryInto;
+use sp_runtime::traits::Hash;
+
+pub mod types;
+pub use types::*;
+
+type String<T> = BoundedVec<u8, <T as pallet::Config>::StringLimit>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+  
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>>
+			+ IsType<<Self as frame_system::Config>::Event>
+			+ Into<<Self as frame_system::Config>::Event>;
+
+		/// The maximum length of a name or cid stored on-chain.
+		#[pallet::constant]
+		type StringLimit: Get<u32>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// New BattlePass created
+		BattlepassCreated {
+			// org_id: T::Hash,
+			org_id: T::Hash,
+			battlepass_id: T::Hash,
+			block_number: T::BlockNumber
+		},
+		
+		/// BattlePass claimed
+		BattlepassClaimed {
+			claimer: T::AccountId,
+			org_id: T::Hash,
+			battlepass_id: T::Hash,
+			nft_id: T::Hash,
+			block_number: T::BlockNumber
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AuthorizationError,
+		MissingParameter,
+		InvalidParameter,
+		OrganizationUnknown,
+		OrgHasActiveBattlepass,
+		BattlepassExists,
+		BattlepassClaimed,
+		BattlepassUnknown,
+		NotMember,
+	  }
+
+	/// Battlepass by its id.
+	///
+	/// Battlepasses: map Hash => Battlepass
+	#[pallet::storage]
+	#[pallet::getter(fn get_battlepass)]
+	pub(super) type Battlepasses<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, Battlepass<T::Hash, T::AccountId, T::BlockNumber, String<T>>, OptionQuery>;
+
+	/// Number of battlepasses per organization.
+	///
+	/// BattlepassCount: map Hash => u32
+	#[pallet::storage]
+	#[pallet::getter(fn get_battlepass_count)]
+	pub type BattlepassCount<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, u32, ValueQuery>;
+
+	/// Current active battlepass in organization.
+	///
+	/// ActiveBattlepassByOrg: map Hash => Hash
+	#[pallet::storage]
+	#[pallet::getter(fn get_active_battlepass)]
+	pub type ActiveBattlepassByOrg<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, T::Hash, OptionQuery>;
+
+	/// Claimed Battlepass by user.
+	///
+	/// ClaimedBattlepass: map (AccountId, Hash) => Hash
+	#[pallet::storage]
+	#[pallet::getter(fn get_claimed_battlepass)]
+	pub(super) type ClaimedBattlepass<T: Config> = StorageDoubleMap<_,
+		Blake2_128Concat, T::AccountId,
+		Blake2_128Concat, T::Hash,
+		T::Hash,
+		OptionQuery
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn create_battlepass(
```

Comment:
if you put the transactional part into the extrinsic fn, you may need to double the code unneccessarily as we need a way to add a battlepass via oracle, therefore `who, pass, account` where who is permitted oracle and account is the user to add. also we need to store respective discord id somewhere otherwise we cannot map it,.

Reply:
I think there is slight confusion in terminology.
We name two entities the same way: `a)` general entity, which is created by organization and represents event to participate for the users; `b)` entity, which users purchase to gain access to participate in the event.
Particularly this extrinsic is used to create `a)` entity type. So we don't need to store discord id here and extend functionality for the oracle. Those will be applied for working with `b)` entity type.

Reply:
battlepass is the general seasonal wrapper for achievements which result in scores and which unlock claiming rewards. i have written a doc for it

Reply:
pls check here @vayesy @FiberMan @vovacha https://github.com/gamedaoco/gamedao-protocol/pull/120/files#diff-46df5e399469d42e3bf99a2d701851c5c5ca55ac2897b4a0bedf423ef03d94ae

Reply:
can you add the proper names for those entities and structs and show where they are stored? @vayesy